plugin.program.remote.control.browser/plugin.py

```diff
@@ -0,0 +1,990 @@
+                terminator = threading.Timer(
+                    BROWSER_EXIT_DELAY.total_seconds(), terminateBrowser)
+                terminator.start()
+                try:
+                    proc.wait()
+                    proc = None
+                    xbmc.log('Waited for the browser to quit', xbmc.LOGDEBUG)
+                finally:
+                    terminator.cancel()
+                    terminator.join()
+            finally:
+                if proc is not None:
+                    # As a last resort, forcibly kill the browser.
+                    xbmc.log('Forcefully killing the browser', xbmc.LOGINFO)
+                    killBrowser(proc, signal.SIGKILL)
+                    proc.wait()
+                    xbmc.log('Waited for the browser to die', xbmc.LOGDEBUG)
+        finally:
+            if waiter is not None:
+                xbmc.log(
+                    'Joining with browser monitoring thread', xbmc.LOGDEBUG)
+                waiter.join()
+                xbmc.log(
+                    'Joined with browser monitoring thread', xbmc.LOGDEBUG)
+
+
+def activateWindow(cmd, proc, isAborting, xdotoolPath):
+    (output, error) = proc.communicate()
+    if isAborting.is_set():
+        xbmc.log('Aborting search for browser PID', xbmc.LOGDEBUG)
+        return
+    if proc.returncode != 0:
+        xbmc.log('Failed search for browser PID', xbmc.LOGINFO)
+        raise subprocess.CalledProcessError(proc.returncode, cmd, output)
+    wids = [wid for wid in output.split('\n') if wid]
+    for wid in wids:
+        if isAborting.is_set():
+            xbmc.log('Aborting activation of windows', xbmc.LOGDEBUG)
+            return
+        xbmc.log('Activating window with WID: ' + wid, xbmc.LOGDEBUG)
+        subprocess.call([xdotoolPath, 'WindowActivate', wid])
+    xbmc.log('Finished activating windows', xbmc.LOGDEBUG)
+
+
+@contextlib.contextmanager
+def raiseBrowser(pid, xdotoolPath):
+    if not xdotoolPath:
+        xbmc.log('Not raising the browser')
+        yield
+        return
+    activator = None
+    # With the "sync" flag, the command could block indefinitely.
+    cmd = [xdotoolPath, 'search', '--sync', '--onlyvisible', '--pid', str(pid)]
+    xbmc.log(
+        'Searching for browser PID: ' +
+        ' '.join(pipes.quote(arg) for arg in cmd),
+        xbmc.LOGINFO)
+    proc = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, universal_newlines=True)
+    try:
+        isAborting = threading.Event()
+        activatorStarting = threading.Thread(
+            target=activateWindow, args=(cmd, proc, isAborting, xdotoolPath))
+        activatorStarting.start()
+        activator = activatorStarting
+
+        yield
+    finally:
+        isAborting.set()
+        if proc.poll() is None:
+            proc.kill()
+        proc.wait()
+        if activator is not None:
+            activator.join()
+
+
+@contextlib.contextmanager
+def lockPidfile(browserLockPath, pid):
```

Comment:
The user can trigger a browser to start either though the Kodi GUI or through the remote bookmarklet API. But with multiple start mechanisms, they need to make sure they don't clobber each other. If the plugin is already running, then other threads that try to start the browser will just fail. The pidfile is the locking mechanism.